tests/test_TFEngine.py

```diff
@@ -1163,6 +1163,126 @@ def test_attention_two_dependent_targets():
+
+  config = Config()
+  config.update({
+    "model": "%s/model" % _get_tmp_dir(),
+    "batch_size": 5000,
+    "max_seqs": 2,
+    "extern_data": {"data": {"dim": n_data_dim, "sparse": False},
+      "classes_0": {"dim": n_classes_dim_0, "sparse": True, "same_dim_tags_as": {"t": dec_time}},
+      "classes_1": {"dim": n_classes_dim_1, "sparse": True, "same_dim_tags_as": {"t": dec_time}},
+      "classes_2": {"dim": n_classes_dim_2, "sparse": True, "same_dim_tags_as": {"t": dec_time}},
+    },
+    "num_epochs": 1,
+    "network": net_dict,
+    "debug_print_layer_output_template": True,
+  })
+  _cleanup_old_models(config)
+  engine = Engine(config=config)
+  print("Train...")
+  engine.init_train_from_config(config=config, train_data=data, dev_data=None)
+  engine.train()
+
+  print("Search...")
+  engine.use_search_flag = True
+  engine.use_dynamic_train_flag = False
+  engine.init_network_from_config(config)
+  engine.search(dataset=data, output_layer_names=["decision_0", "decision_1", "decision_2"])
+  assert engine.network.total_objective is not None
+  assert "decision_0" in engine.network.losses_dict
+  assert "decision_1" in engine.network.losses_dict
+  assert "decision_2" in engine.network.losses_dict
+
+  engine.finalize()
+
+
+def test_attention_three_dependent_targets():
+  """
+  Tests training and search when having three ChoiceLayers in the loop that depend on each other.
+  Note, there will be different beam sizes in different parts of the recurrent unit.
+  """
+  beam_size_0 = 5
+  beam_size_1 = 3
+  beam_size_2 = 7
+
+  net_dict = {
+    "encoder": {"class": "linear", "activation": "tanh", "n_out": 5},
+    "output": {
+      "class": "rec",
+      "from": [],
+      "target": "classes_2", "max_seq_len": 10,
+      "unit": {
+        "end": {"class": "compare", "from": ["output_0"], "value": 0},
+        "orth_embed_0": {'class': 'linear', 'activation': None, 'from': ['output_0'], "n_out": 7},
+        "orth_embed_1": {'class': 'linear', 'activation': None, 'from': ['output_1'], "n_out": 7},
+        "orth_embed_2": {'class': 'linear', 'activation': None, 'from': ['output_2'], "n_out": 7},
+        "orth_embed": {"class": "copy", "from": ["orth_embed_0", "orth_embed_1", "orth_embed_2"]},
+        "s": {"class": "rnn_cell", "unit": "LSTMBlock", "from": ["prev:c", "prev:orth_embed"], "n_out": 7},
+        "c_in": {"class": "linear", "activation": "tanh", "from": ["s", "prev:orth_embed"], "n_out": 5},
+        "c": {"class": "dot_attention", "from": ["c_in"], "base": "base:encoder", "base_ctx": "base:encoder"},
+        "output_prob_0": {"class": "softmax", "from": ["prev:s", "c"], "target": "classes_0", "loss": "ce"},
+        "output_prob_1": {"class": "softmax", "from": ["prev:s", "c", "orth_embed_0"],
+          "target": "classes_1", "loss": "ce"},
+        "output_prob_1": {"class": "softmax", "from": ["prev:s", "c", "orth_embed_0", "orth_embed_1"],
```

Comment:
Oops, this should be `output_prob_2`. With this change it runs. But maybe this only means this test does not reproduce my error, looking into it...

Reply:
This test is also not really about multiple targets (and I don't remember what that means anyway), but just about multiple choices.